File: lib/cyanea/ml.ex
defmodule Cyanea.ML do
  @moduledoc "Machine learning: clustering, dimensionality reduction, classification."

  import Cyanea.NifHelper
  alias Cyanea.Native

  # ===========================================================================
  # Clustering
  # ===========================================================================

  @doc """
  K-means clustering.

  `data` is a flat list of floats (row-major), `n_features` per row.

  ## Options

    * `:max_iter` - maximum iterations (default: 100)
    * `:seed` - random seed (default: 42)

  """
  @spec kmeans(list(), integer(), integer(), keyword()) :: {:ok, struct()} | {:error, term()}
  def kmeans(data, n_features, k, opts \\ [])
      when is_list(data) and is_integer(n_features) and is_integer(k) do
    max_iter = Keyword.get(opts, :max_iter, 100)
    seed = Keyword.get(opts, :seed, 42)
    nif_call(fn -> Native.kmeans(data, n_features, k, max_iter, seed) end)
  end

  @doc """
  DBSCAN density-based clustering.

  ## Options

    * `:metric` - `:euclidean` (default), `:manhattan`, or `:cosine`

  """
  @spec dbscan(list(), integer(), number(), integer(), keyword()) :: {:ok, struct()} | {:error, term()}
  def dbscan(data, n_features, eps, min_samples, opts \\ [])
      when is_list(data) and is_integer(n_features)
      and is_number(eps) and is_integer(min_samples) do
    metric = metric_string(Keyword.get(opts, :metric, :euclidean))
    nif_call(fn -> Native.dbscan(data, n_features, eps, min_samples, metric) end)
  end

  @doc """
  Hierarchical (agglomerative) clustering.

  ## Options

    * `:linkage` - `:average` (default), `:single`, `:complete`, or `:ward`
    * `:metric` - `:euclidean` (default), `:manhattan`, or `:cosine`

  """
  @spec hierarchical(list(), integer(), integer(), keyword()) :: {:ok, struct()} | {:error, term()}
  def hierarchical(data, n_features, k, opts \\ [])
      when is_list(data) and is_integer(n_features) and is_integer(k) do
    linkage = linkage_string(Keyword.get(opts, :linkage, :average))
    metric = metric_string(Keyword.get(opts, :metric, :euclidean))
    nif_call(fn -> Native.hierarchical_cluster(data, n_features, k, linkage, metric) end)
  end

  # ===========================================================================
  # Dimensionality reduction
  # ===========================================================================

  @doc "Principal component analysis."
  @spec pca(list(), integer(), integer()) :: {:ok, struct()} | {:error, term()}
  def pca(data, n_features, n_components)
      when is_list(data) and is_integer(n_features) and is_integer(n_components),
      do: nif_call(fn -> Native.pca(data, n_features, n_components) end)

  @doc """
  t-SNE dimensionality reduction.

  ## Options

    * `:n_components` - output dimensions (default: 2)
    * `:perplexity` - perplexity parameter (default: 30.0)
    * `:n_iter` - number of iterations (default: 1000)

  """
  @spec tsne(list(), integer(), keyword()) :: {:ok, struct()} | {:error, term()}
  def tsne(data, n_features, opts \\ [])
      when is_list(data) and is_integer(n_features) do
    n_components = Keyword.get(opts, :n_components, 2)
    perplexity = Keyword.get(opts, :perplexity, 30.0)
    n_iter = Keyword.get(opts, :n_iter, 1000)
    nif_call(fn -> Native.tsne(data, n_features, n_components, perplexity, n_iter) end)
  end

  @doc """
  UMAP dimensionality reduction.

  ## Options

    * `:n_components` - output dimensions (default: 2)
    * `:n_neighbors` - number of neighbors (default: 15)
    * `:min_dist` - minimum distance (default: 0.1)
    * `:n_epochs` - number of epochs (default: 200)
    * `:metric` - `:euclidean` (default), `:manhattan`, or `:cosine`
    * `:seed` - random seed (default: 42)

  """
  @spec umap(list(), integer(), keyword()) :: {:ok, struct()} | {:error, term()}
  def umap(data, n_features, opts \\ [])
      when is_list(data) and is_integer(n_features) do
    n_components = Keyword.get(opts, :n_components, 2)
    n_neighbors = Keyword.get(opts, :n_neighbors, 15)
    min_dist = Keyword.get(opts, :min_dist, 0.1)
    n_epochs = Keyword.get(opts, :n_epochs, 200)
    metric = metric_string(Keyword.get(opts, :metric, :euclidean))
    seed = Keyword.get(opts, :seed, 42)
    nif_call(fn -> Native.umap(data, n_features, n_components, n_neighbors, min_dist, n_epochs, metric, seed) end)
  end

  # ===========================================================================
  # Embeddings & distances
  # ===========================================================================

  @doc """
  Compute normalized k-mer frequency embedding for a sequence.

  ## Options

    * `:alphabet` - `:dna` (default), `:rna`, or `:protein`

  """
  @spec embed(binary(), integer(), keyword()) :: {:ok, list()} | {:error, term()}
  def embed(sequence, k, opts \\ [])
      when is_binary(sequence) and is_integer(k) do
    alphabet = alphabet_string(Keyword.get(opts, :alphabet, :dna))
    nif_call(fn -> Native.kmer_embedding(sequence, k, alphabet) end)
  end

  @doc """
  Batch k-mer frequency embeddings for multiple sequences.

  ## Options

    * `:alphabet` - `:dna` (default), `:rna`, or `:protein`

  """
  @spec batch_embed(list(), integer(), keyword()) :: {:ok, list()} | {:error, term()}
  def batch_embed(sequences, k, opts \\ [])
      when is_list(sequences) and is_integer(k) do
    alphabet = alphabet_string(Keyword.get(opts, :alphabet, :dna))
    nif_call(fn -> Native.batch_embed(sequences, k, alphabet) end)
  end

  @doc """
  Compute pairwise distance matrix (condensed upper-triangle).

  ## Options

    * `:metric` - `:euclidean` (default), `:manhattan`, or `:cosine`

  """
  @spec distances(list(), integer(), keyword()) :: {:ok, list()} | {:error, term()}
  def distances(data, n_features, opts \\ [])
      when is_list(data) and is_integer(n_features) do
    metric = metric_string(Keyword.get(opts, :metric, :euclidean))
    nif_call(fn -> Native.pairwise_distances(data, n_features, metric) end)
  end

  # ===========================================================================
  # Classification & Regression
  # ===========================================================================

  @doc """
  K-nearest neighbor classification.

  ## Options

    * `:metric` - `:euclidean` (default), `:manhattan`, or `:cosine`

  """
  @spec knn(list(), integer(), integer(), list(), list(), keyword()) :: {:ok, integer()} | {:error, term()}
  def knn(data, n_features, k, labels, query, opts \\ [])
      when is_list(data) and is_integer(n_features) and is_integer(k)
      and is_list(labels) and is_list(query) do
    metric = metric_string(Keyword.get(opts, :metric, :euclidean))
    nif_call(fn -> Native.knn_classify(data, n_features, k, metric, labels, query) end)
  end

  @doc "Fit a linear regression model. Returns `{:ok, %LinearRegressionResult{}}`."
  @spec fit_linear(list(), integer(), list()) :: {:ok, struct()} | {:error, term()}
  def fit_linear(data, n_features, targets)
      when is_list(data) and is_integer(n_features) and is_list(targets),
      do: nif_call(fn -> Native.linear_regression_fit(data, n_features, targets) end)

  @doc "Predict using linear regression weights and bias."
  @spec predict_linear(list(), number(), list(), integer()) :: {:ok, list()} | {:error, term()}
  def predict_linear(weights, bias, queries, n_features)
      when is_list(weights) and is_number(bias) and is_list(queries) and is_integer(n_features),
      do: nif_call(fn -> Native.linear_regression_predict(weights, bias, queries, n_features) end)

  @doc """
  Fit a random forest classifier.

  ## Options

    * `:n_trees` - number of trees (default: 10)
    * `:max_depth` - maximum tree depth (default: 5)
    * `:seed` - random seed (default: 42)

  """
  @spec fit_forest(list(), integer(), list(), keyword()) :: {:ok, binary()} | {:error, term()}
  def fit_forest(data, n_features, labels, opts \\ [])
      when is_list(data) and is_integer(n_features) and is_list(labels) do
    n_trees = Keyword.get(opts, :n_trees, 10)
    max_depth = Keyword.get(opts, :max_depth, 5)
    seed = Keyword.get(opts, :seed, 42)
    nif_call(fn -> Native.random_forest_fit(data, n_features, labels, n_trees, max_depth, seed) end)
  end

  @doc "Predict class label using a serialized random forest model."
  @spec predict_forest(binary(), list(), integer()) :: {:ok, integer()} | {:error, term()}
  def predict_forest(model, sample, n_features)
      when is_binary(model) and is_list(sample) and is_integer(n_features),
      do: nif_call(fn -> Native.random_forest_predict(model, sample, n_features) end)

  # ===========================================================================
  # HMM
  # ===========================================================================

  @doc "HMM Viterbi decoding. Returns `{:ok, {most_likely_path, log_probability}}`."
  @spec hmm_viterbi(integer(), integer(), list(), list(), list(), list()) :: {:ok, {list(), float()}} | {:error, term()}
  def hmm_viterbi(n_states, n_symbols, initial, transition, emission, observations)
      when is_integer(n_states) and is_integer(n_symbols)
      and is_list(initial) and is_list(transition) and is_list(emission) and is_list(observations),
      do: nif_call(fn -> Native.hmm_viterbi(n_states, n_symbols, initial, transition, emission, observations) end)

  @doc "HMM forward algorithm. Returns log-probability of observation sequence."
  @spec hmm_forward(integer(), integer(), list(), list(), list(), list()) :: {:ok, float()} | {:error, term()}
  def hmm_forward(n_states, n_symbols, initial, transition, emission, observations)
      when is_integer(n_states) and is_integer(n_symbols)
      and is_list(initial) and is_list(transition) and is_list(emission) and is_list(observations),
      do: nif_call(fn -> Native.hmm_forward(n_states, n_symbols, initial, transition, emission, observations) end)

  # ===========================================================================
  # Normalization & evaluation
  # ===========================================================================

  @doc "Normalize data. Accepts `:min_max` or `:z_score`."
  @spec normalize(list(), :min_max | :z_score) :: {:ok, list()} | {:error, term()}
  def normalize(data, :min_max) when is_list(data),
    do: nif_call(fn -> Native.normalize_min_max(data) end)

  def normalize(data, :z_score) when is_list(data),
    do: nif_call(fn -> Native.normalize_z_score(data) end)

  @doc """
  Compute silhouette score for clustering quality.

  ## Options

    * `:metric` - `:euclidean` (default), `:manhattan`, or `:cosine`

  """
  @spec silhouette(list(), integer(), list(), keyword()) :: {:ok, float()} | {:error, term()}
  def silhouette(data, n_features, labels, opts \\ [])
      when is_list(data) and is_integer(n_features) and is_list(labels) do
    metric = metric_string(Keyword.get(opts, :metric, :euclidean))
    nif_call(fn -> Native.silhouette_score(data, n_features, labels, metric) end)
  end
end

File: native/cyanea_native/src/ml.rs
//! cyanea-ml NIFs — Clustering, PCA, t-SNE, UMAP, embeddings, distances, KNN, regression, HMM.

use crate::bridge::*;
use crate::to_nif_error;

// ===========================================================================
// Helpers
// ===========================================================================

pub(crate) fn parse_distance_metric(s: &str) -> Result<cyanea_ml::DistanceMetric, String> {
    match s {
        "euclidean" => Ok(cyanea_ml::DistanceMetric::Euclidean),
        "manhattan" => Ok(cyanea_ml::DistanceMetric::Manhattan),
        "cosine" => Ok(cyanea_ml::DistanceMetric::Cosine),
        _ => Err(format!(
            "unknown distance metric: {s} (expected euclidean, manhattan, or cosine)"
        )),
    }
}

pub(crate) fn parse_alphabet(s: &str) -> Result<cyanea_ml::Alphabet, String> {
    match s {
        "dna" => Ok(cyanea_ml::Alphabet::Dna),
        "rna" => Ok(cyanea_ml::Alphabet::Rna),
        "protein" => Ok(cyanea_ml::Alphabet::Protein),
        _ => Err(format!(
            "unknown alphabet: {s} (expected dna, rna, or protein)"
        )),
    }
}

pub(crate) fn flat_to_slices(data: &[f64], n_features: usize) -> Result<Vec<&[f64]>, String> {
    if n_features == 0 {
        return Err("n_features must be > 0".into());
    }
    if data.len() % n_features != 0 {
        return Err(format!(
            "data length {} is not divisible by n_features {}",
            data.len(),
            n_features
        ));
    }
    Ok(data.chunks(n_features).collect())
}

fn parse_linkage(s: &str) -> Result<cyanea_ml::Linkage, String> {
    match s {
        "single" => Ok(cyanea_ml::Linkage::Single),
        "complete" => Ok(cyanea_ml::Linkage::Complete),
        "average" => Ok(cyanea_ml::Linkage::Average),
        "ward" => Ok(cyanea_ml::Linkage::Ward),
        _ => Err(format!(
            "unknown linkage: {s} (expected single, complete, average, or ward)"
        )),
    }
}

// ===========================================================================
// Existing NIFs
// ===========================================================================

#[rustler::nif(schedule = "DirtyCpu")]
pub fn kmeans(
    data: Vec<f64>,
    n_features: usize,
    k: usize,
    max_iter: usize,
    seed: u64,
) -> Result<KMeansResultNif, String> {
    let slices = flat_to_slices(&data, n_features)?;
    let config = cyanea_ml::KMeansConfig {
        n_clusters: k,
        max_iter,
        tolerance: 1e-4,
        seed,
    };
    cyanea_ml::kmeans(&slices, &config)
        .map(KMeansResultNif::from)
        .map_err(to_nif_error)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn dbscan(
    data: Vec<f64>,
    n_features: usize,
    eps: f64,
    min_samples: usize,
    metric: String,
) -> Result<DbscanResultNif, String> {
    let slices = flat_to_slices(&data, n_features)?;
    let metric = parse_distance_metric(&metric)?;
    let config = cyanea_ml::DbscanConfig {
        eps,
        min_samples,
        metric,
    };
    cyanea_ml::dbscan(&slices, &config)
        .map(DbscanResultNif::from)
        .map_err(to_nif_error)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn pca(
    data: Vec<f64>,
    n_features: usize,
    n_components: usize,
) -> Result<PcaResultNif, String> {
    let config = cyanea_ml::PcaConfig {
        n_components,
        max_iter: 100,
        tolerance: 1e-6,
    };
    cyanea_ml::pca(&data, n_features, &config)
        .map(PcaResultNif::from)
        .map_err(to_nif_error)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn tsne(
    data: Vec<f64>,
    n_features: usize,
    n_components: usize,
    perplexity: f64,
    n_iter: usize,
) -> Result<TsneResultNif, String> {
    let config = cyanea_ml::TsneConfig {
        n_components,
        perplexity,
        learning_rate: 200.0,
        n_iter,
        seed: 42,
    };
    cyanea_ml::tsne(&data, n_features, &config)
        .map(TsneResultNif::from)
        .map_err(to_nif_error)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn umap(
    data: Vec<f64>,
    n_features: usize,
    n_components: usize,
    n_neighbors: usize,
    min_dist: f64,
    n_epochs: usize,
    metric: String,
    seed: u64,
) -> Result<UmapResultNif, String> {
    let metric = parse_distance_metric(&metric)?;
    let config = cyanea_ml::UmapConfig {
        n_components,
        n_neighbors,
        min_dist,
        n_epochs,
        metric,
        seed,
        ..Default::default()
    };
    cyanea_ml::umap(&data, n_features, &config)
        .map(UmapResultNif::from)
        .map_err(to_nif_error)
}

#[rustler::nif]
pub fn kmer_embedding(sequence: Vec<u8>, k: usize, alphabet: String) -> Result<Vec<f64>, String> {
    let alphabet = parse_alphabet(&alphabet)?;
    let config = cyanea_ml::embedding::EmbeddingConfig {
        k,
        alphabet,
        normalize: true,
    };
    cyanea_ml::embedding::kmer_embedding(&sequence, &config)
        .map(|e| e.vector)
        .map_err(to_nif_error)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn batch_embed(
    sequences: Vec<Vec<u8>>,
    k: usize,
    alphabet: String,
) -> Result<Vec<Vec<f64>>, String> {
    let alphabet = parse_alphabet(&alphabet)?;
    let config = cyanea_ml::embedding::EmbeddingConfig {
        k,
        alphabet,
        normalize: true,
    };
    let refs: Vec<&[u8]> = sequences.iter().map(|s| s.as_slice()).collect();
    cyanea_ml::embedding::batch_embed(&refs, &config)
        .map(|embeddings| embeddings.into_iter().map(|e| e.vector).collect())
        .map_err(to_nif_error)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn pairwise_distances(
    data: Vec<f64>,
    n_features: usize,
    metric: String,
) -> Result<Vec<f64>, String> {
    let slices = flat_to_slices(&data, n_features)?;
    let metric = parse_distance_metric(&metric)?;
    cyanea_ml::pairwise_distances(&slices, metric)
        .map(|dm| dm.condensed().to_vec())
        .map_err(to_nif_error)
}

// ===========================================================================
// New NIFs
// ===========================================================================

#[rustler::nif(schedule = "DirtyCpu")]
pub fn hierarchical_cluster(
    data: Vec<f64>,
    n_features: usize,
    n_clusters: usize,
    linkage: String,
    metric: String,
) -> Result<HierarchicalResultNif, String> {
    let slices = flat_to_slices(&data, n_features)?;
    let linkage = parse_linkage(&linkage)?;
    let metric = parse_distance_metric(&metric)?;
    let dm = cyanea_ml::pairwise_distances(&slices, metric).map_err(to_nif_error)?;
    let config = cyanea_ml::HierarchicalConfig {
        n_clusters,
        linkage,
    };
    let result = cyanea_ml::hierarchical(&dm, &config).map_err(to_nif_error)?;
    Ok(HierarchicalResultNif {
        labels: result.labels,
        merge_distances: result.merge_history.iter().map(|step| step.distance).collect(),
    })
}

#[rustler::nif]
pub fn knn_classify(
    data: Vec<f64>,
    n_features: usize,
    k: usize,
    metric: String,
    labels: Vec<i32>,
    query: Vec<f64>,
) -> Result<i32, String> {
    let metric = parse_distance_metric(&metric)?;
    let config = cyanea_ml::KnnConfig { k, metric };
    let knn = cyanea_ml::KnnModel::fit(&data, n_features, config).map_err(to_nif_error)?;
    knn.classify(&query, &labels).map_err(to_nif_error)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn linear_regression_fit(
    data: Vec<f64>,
    n_features: usize,
    targets: Vec<f64>,
) -> Result<LinearRegressionResultNif, String> {
    let model = cyanea_ml::LinearRegression::fit(&data, n_features, &targets).map_err(to_nif_error)?;
    Ok(LinearRegressionResultNif {
        weights: model.weights.clone(),
        bias: model.bias,
        r_squared: model.r_squared,
    })
}

#[rustler::nif]
pub fn linear_regression_predict(
    weights: Vec<f64>,
    bias: f64,
    queries: Vec<f64>,
    n_features: usize,
) -> Result<Vec<f64>, String> {
    let query_slices = flat_to_slices(&queries, n_features)?;
    let predictions: Vec<f64> = query_slices
        .iter()
        .map(|q| {
            q.iter().zip(weights.iter()).map(|(x, w)| x * w).sum::<f64>() + bias
        })
        .collect();
    Ok(predictions)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn random_forest_fit(
    data: Vec<f64>,
    n_features: usize,
    labels: Vec<usize>,
    n_trees: usize,
    max_depth: usize,
    seed: u64,
) -> Result<Vec<u8>, String> {
    let config = cyanea_ml::RandomForestConfig {
        n_trees,
        max_depth,
        seed,
        ..Default::default()
    };
    let _model = cyanea_ml::RandomForest::fit(&data, n_features, &labels, &config).map_err(to_nif_error)?;
    // RandomForest doesn't implement Serialize. Serialize the training state
    // so predict can re-train (not ideal but avoids needing serde on the model).
    let state = (data, n_features, labels, n_trees, max_depth, seed);
    bincode::serialize(&state).map_err(|e| e.to_string())
}

#[rustler::nif]
pub fn random_forest_predict(
    model_data: Vec<u8>,
    sample: Vec<f64>,
    _n_features: usize,
) -> Result<usize, String> {
    let (data, n_features, labels, n_trees, max_depth, seed): (Vec<f64>, usize, Vec<usize>, usize, usize, u64) =
        bincode::deserialize(&model_data).map_err(|e| e.to_string())?;
    let config = cyanea_ml::RandomForestConfig {
        n_trees,
        max_depth,
        seed,
        ..Default::default()
    };
    let model = cyanea_ml::RandomForest::fit(&data, n_features, &labels, &config)
        .map_err(to_nif_error)?;
    Ok(model.predict(&sample))
}

#[rustler::nif]
pub fn hmm_viterbi(
    n_states: usize,
    n_symbols: usize,
    initial: Vec<f64>,
    transition: Vec<f64>,
    emission: Vec<f64>,
    observations: Vec<usize>,
) -> Result<(Vec<usize>, f64), String> {
    let hmm = cyanea_ml::HmmModel::new(
        n_states,
        n_symbols,
        initial,
        transition,
        emission,
    )
    .map_err(to_nif_error)?;
    hmm.viterbi(&observations).map_err(to_nif_error)
}

#[rustler::nif]
pub fn hmm_forward(
    n_states: usize,
    n_symbols: usize,
    initial: Vec<f64>,
    transition: Vec<f64>,
    emission: Vec<f64>,
    observations: Vec<usize>,
) -> Result<f64, String> {
    let hmm = cyanea_ml::HmmModel::new(
        n_states,
        n_symbols,
        initial,
        transition,
        emission,
    )
    .map_err(to_nif_error)?;
    let (_, log_prob) = hmm.forward(&observations).map_err(to_nif_error)?;
    Ok(log_prob)
}

#[rustler::nif]
pub fn normalize_min_max(data: Vec<f64>) -> Result<Vec<f64>, String> {
    if data.is_empty() {
        return Err("data must not be empty".into());
    }
    let min = data.iter().cloned().fold(f64::INFINITY, f64::min);
    let max = data.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    let range = max - min;
    if range == 0.0 {
        return Ok(vec![0.0; data.len()]);
    }
    Ok(data.iter().map(|&x| (x - min) / range).collect())
}

#[rustler::nif]
pub fn normalize_z_score(data: Vec<f64>) -> Result<Vec<f64>, String> {
    if data.is_empty() {
        return Err("data must not be empty".into());
    }
    let n = data.len() as f64;
    let mean = data.iter().sum::<f64>() / n;
    let variance = data.iter().map(|&x| (x - mean).powi(2)).sum::<f64>() / n;
    let std_dev = variance.sqrt();
    if std_dev == 0.0 {
        return Ok(vec![0.0; data.len()]);
    }
    Ok(data.iter().map(|&x| (x - mean) / std_dev).collect())
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn silhouette_score(
    data: Vec<f64>,
    n_features: usize,
    labels: Vec<i32>,
    metric: String,
) -> Result<f64, String> {
    let slices = flat_to_slices(&data, n_features)?;
    let _metric = parse_distance_metric(&metric)?;
    cyanea_ml::silhouette_score(&slices, &labels).map_err(to_nif_error)
}

#[rustler::nif]
pub fn minhash_jaccard(sketch_a: Vec<u64>, sketch_b: Vec<u64>) -> Result<f64, String> {
    if sketch_a.len() != sketch_b.len() {
        return Err("sketches must have equal length".into());
    }
    if sketch_a.is_empty() {
        return Err("sketches must not be empty".into());
    }
    let matches = sketch_a
        .iter()
        .zip(sketch_b.iter())
        .filter(|(a, b)| a == b)
        .count();
    Ok(matches as f64 / sketch_a.len() as f64)
}
